### PR TITLE
Refactor/timeseries to numpy

### PR DESCRIPTION
**Pull request by dennisbader**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

### Summary
- migrate `TimeSeries` backend from `xarray` to `numpy` for huge performance boost
- identical behavior from user-perspective
- the only methods which still rely on xarray are `__repr__` (and `__str__`, `_to_html`), `plot`, and `resample`

### Additional Info

Here are the benchmarks for performance boost per method.

The y-axis shows different TimeSeries shapes (n time steps, n columns, n samples).
The x-axis shows different TimeSeries methods

Explanation: values means (x + 1) times as fast as before. E.g.:

- `0` means same time as before
- `1` means twice as fast as before
- `2` means three times as fast as before

results_general

results_no_copy

Benchmarking script (this requires to have the old `TimeSeries` in a file `darts.timeseries_old`)

```py
import time

import numpy as np
import pandas as pd


def timing_decorator(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()
        for i in range(10000):
            func(*args, **kwargs)
        end_time = time.time()
        execution_time = end_time - start_time
        execution_time = '{0:.6f}'.format(execution_time)
        fname = func.__name__
        print(f"Function '{func.__name__}' executed in {execution_time} seconds")
        return fname, execution_time
    return wrapper


@timing_decorator
def from_times_and_values_copy(idx, vals):
    _ = TimeSeries.from_times_and_values(
        times=idx,
        values=vals,
    )


@timing_decorator
def from_times_and_values_no_copy(idx, vals):
    _ = TimeSeries(
        times=idx,
        values=vals,
        copy=False,
    )


@timing_decorator
def getitem_slice(series):
    _ = series[3::2]


@timing_decorator
def getitem_slice_ts(series):
    tidx = series._time_index
    start, end, step = tidx[3], tidx[-2], 2
    _ = series[start:end:step]


@timing_decorator
def arithmetics(series):
    _ = series + 2.


# run once for new TimeSeries (`True`), once for old TimeSeries (`False`)
use_new = True
if use_new:
    from darts import TimeSeries
else:
    from darts.timeseries_old import TimeSeries


if __name__ == "__main__":
    # benchmark series with different shapes
    n_sets = [1000, 100, 10]  # number of time steps
    n_cols_sets = [10, 5, 1]  # number of columns
    n_samples_sets = [1000, 100, 1]  # number of samples

    results = {}
    index = []
    for n in n_sets:
        for n_cols in n_cols_sets:
            for n_samples in n_samples_sets:
                # sorted index (int, range, time)
                idx1 = pd.Index([i for i in range(n)])
                idx2 = pd.RangeIndex(start=0, stop=n, step=1)
                idx3 = pd.date_range("2000-01-01", periods=n, freq="D")

                # shuffled index (int, time)
                idx4 = pd.Index(np.concatenate([idx1[:2].values, idx1[2:][::-1].values]))
                idx5 = pd.DatetimeIndex(np.concatenate([idx2[:2].values, idx2[2:][::-1].values]))

                values = np.empty((n, n_cols, n_samples))
                shape = str(values.shape)
                print(shape)

                # TimeSeries for manipulation benchmarks
                series_iidx = TimeSeries.from_times_and_values(times=idx2, values=values)
                series_tidx = TimeSeries.from_times_and_values(times=idx3, values=values)

                indexes = [idx1, idx2, idx3, idx4, idx5]
                names = ["Int Idx", "Range Idx", "Datetime Idx", "Int Idx (shuffled)", "Datetime Idx (shuffled)"]
                series_sets = [series_iidx, series_tidx]
                series_names = ["Series Range Idx", "Series Datetime Idx"]

                # series creation benchmarks
                for idx_, name in zip(indexes, names):
                    print(name)
                    f_name, t_ex = from_times_and_values_copy(idx_, values)
                    results["_".join([f_name, name, shape])] = [t_ex]
                    if use_new:
                        f_name, t_ex = from_times_and_values_no_copy(idx_, values)
                        results["_".join([f_name, name, shape])] = [t_ex]

                # series manipulation benchmarks
                for series_, name in zip(series_sets, series_names):
                    print(name)
                    f_name, t_ex = getitem_slice(series_)
                    results["_".join([f_name, name, shape])] = [t_ex]
                    f_name, t_ex = getitem_slice_ts(series_)
                    results["_".join([f_name, name, shape])] = [t_ex]
                    f_name, t_ex = arithmetics(series_)
                    results["_".join([f_name, name, shape])] = [t_ex]

    pd.DataFrame(results, index=["time"]).T.to_csv(f"benchmark_np_ts_{'new' if use_new else 'old'}.csv")
```

### CODE REVIEW

File: darts/timeseries.py [hunk 1]
```diff
@@ -78,9 +73,12 @@
...
 logger = get_logger(__name__)
 
-# dimension names in the DataArray
+# dimension names in the array
 # the "time" one can be different, if it has a name in the underlying Series/DataFrame.
 DIMS = ("time", "component", "sample")
+TIME_AX = 0
+COMP_AX = 1
+SMPL_AX = 2
 AXES = {"time": 0, "component": 1, "sample": 2}
```
Comment (madtoinou) [inline darts/timeseries.py]: Why not use the `TIME_AX`, `COMP_AX` and `SML_AX` variables here?

File: darts/timeseries.py [hunk 2]
```diff
@@ -89,11 +87,26 @@
 HIERARCHY_TAG = "hierarchy"
 METADATA_TAG = "metadata"
 
+# TODO:
+# - check mutability
```
Comment (madtoinou) [inline darts/timeseries.py]: We should add this to the tests

File: darts/timeseries.py [hunk 3]
```diff
@@ -89,11 +87,26 @@
...
+        times: Union[pd.DatetimeIndex, pd.RangeIndex, pd.Index],
+        values: np.ndarray,
+        fill_missing_dates: Optional[bool] = False,
+        freq: Optional[Union[str, int]] = None,
+        components: Optional[pd._typing.Axes] = None,
+        fillna_value: Optional[float] = None,
+        static_covariates: Optional[Union[pd.Series, pd.DataFrame]] = None,
+        hierarchy: Optional[dict] = None,
+        metadata: Optional[dict] = None,
+        copy=True,
```
Comment (madtoinou) [inline darts/timeseries.py]:
```suggestion
        copy: bool =True,
```

File: darts/timeseries.py [hunk 4]
```diff
@@ -108,100 +121,241 @@ def __init__(self, xa: xr.DataArray, copy=True):
...
-                ValueError("The time series must contain numeric values only."), logger
+        # BUGFIX : force time-index to be timezone naive as xarray doesn't support it
+        if isinstance(times, pd.DatetimeIndex) and times.tz is not None:
+            logger.warning(
+                "The `times` argument was associated with a timezone, which is currently not supported "
+                "by xarray. To avoid unexpected behaviour, the tz information was removed. Consider calling "
+                f"`ts.time_index.tz_localize({times.tz})` when exporting the results."
+                "To plot the series with the right time steps, consider setting the matplotlib.pyplot "
+                "`rcParams['timezone']` parameter to automatically convert the time axis back to the "
+                "original timezone."
```
Comment (madtoinou) [inline darts/timeseries.py]: the comment should be changed to remove mentions to `xarray`

File: darts/timeseries.py [hunk 5]
```diff
@@ -108,100 +121,241 @@ def __init__(self, xa: xr.DataArray, copy=True):
...
 
-        # check that columns/component names are unique
-        components = xa.get_index(DIMS[1])
-        if not len(set(components)) == len(components):
+        if not len(times) == len(values):
             raise_log(
-                ValueError(
-                    f"The components (columns) names must be unique. Provided: {components}"
-                ),
+                ValueError("The time index and values must have to same length."),
```
Comment (madtoinou) [inline darts/timeseries.py]:
```suggestion
                ValueError("The time index and values must have the same length."),
```

File: darts/timeseries.py [hunk 6]
```diff
@@ -108,100 +121,241 @@ def __init__(self, xa: xr.DataArray, copy=True):
...
+            has_datetime_index and times.freq is not None
+        ) or has_range_index
 
-        if self._has_datetime_index:
-            # store original freq (see bug of sortby() above).
-            freq_tmp = xa.get_index(self._time_dim).freq
+        if has_datetime_index:
+            time_zone = times.tz
+            if time_zone is not None:
+                # TODO: potential to use time zone-aware index since `TimeSeries` was refactored
```
Comment (madtoinou) [inline darts/timeseries.py]: Do we want to include it in this PR or keep it in a separate issue/PR?

File: darts/__init__.py [hunk 7]
```diff
@@ -42,4 +42,4 @@
 if os.getenv("DARTS_CONFIGURE_MATPLOTLIB", "1") != "0":
     mpl.rcParams.update(u8plots_mplstyle)
 
-__all__ = ["TimeSeries", "concatenate", "slice_intersect"]
+__all__ = ["TimeSeries", "TimeSeriesNP", "concatenate", "slice_intersect"]
```
Comment (madtoinou) [inline darts/__init__.py]: Leftovers of the refactoring to be reverted

File: darts/models/forecasting/kalman_forecaster.py [hunk 8]
```diff
@@ -163,13 +163,12 @@ def _predict(
...
-            columns=self.training_series.columns,
-            static_covariates=self.training_series.static_covariates,
-            hierarchy=self.training_series.hierarchy,
-            metadata=self.training_series.metadata,
+        series_future = TimeSeries(
+            times=time_index,
+            values=placeholder_vals,
+            components=self.training_series.columns,
+            copy=False,
+            **self.training_series._attrs,
```
Comment (madtoinou) [inline darts/models/forecasting/kalman_forecaster.py:171]: Nice!

File: darts/tests/dataprocessing/dtw/test_dtw.py [hunk 9]
```diff
@@ -61,6 +63,9 @@ def test_shift(self):
         series1 = _series_from_values(input1)
         series2 = _series_from_values(input2)
 
+        series1_copy = series1.copy()
+        series2_copy = series2.copy()
```
Comment (madtoinou) [inline darts/tests/dataprocessing/dtw/test_dtw.py:65]: Maybe use a different variable name as it is already used lines 27 and 28? Or maybe, just delete lines 27 and 28 since they are not necessary?

File: darts/tests/test_timeseries.py [hunk 10]
```diff
@@ -43,43 +43,27 @@ def test_creation(self):
...
-        with pytest.raises(ValueError):
-            # duplicated column names
-            ar3 = xr.DataArray(
-                np.random.randn(10, 2, 1),
-                dims=("time", "component", "sample"),
-                coords={"time": self.times, "component": ["a", "a"]},
-                name="time series",
-            )
-            _ = TimeSeries(ar3)
+        # duplicated column names renames the columns
```
Comment (madtoinou) [inline darts/tests/test_timeseries.py:60]: This should be included in the changelog

File: darts/tests/test_timeseries.py [hunk 11]
```diff
@@ -606,283 +475,33 @@ def test_rescale(self):
...
-        TestTimeSeries.helper_test_drop(self, self.series1)
+        helper_test_drop(self.series1)
 
     @pytest.mark.parametrize(
         "config", itertools.product(["D", "2D", 1, 2], [False, True])
     )
     def test_intersect(self, config):
         """Tests slice intersection between two series with datetime or range index with identical and
         mixed frequencies."""
+        print(config)
```
Comment (madtoinou) [inline darts/tests/test_timeseries.py]: To be removed

File: darts/tests/test_timeseries.py [hunk 12]
```diff
@@ -2134,6 +1787,458 @@ def test_schema(self, config):
...
+
+    # end timestamp after series
+    seriesC = test_series.slice(pd.Timestamp("20130108"), pd.Timestamp("20130201"))
+    assert seriesC.start_time() == pd.Timestamp("20130108")
+    assert seriesC.end_time() == pd.Timestamp("20130110")
+
+    # integer-indexed series, starting at 0
+    values = np.random.rand(30)
+    idx = pd.RangeIndex(start=0, stop=30, step=1)
+    ts = TimeSeries.from_times_and_values(idx, values)
```
Comment (madtoinou) [inline darts/tests/test_timeseries.py]: Shouldn't this be converted to `TimeSeries(...)` instead of using the factory method?

File: darts/tests/test_timeseries_static_covariates.py [hunk 13]
```diff
@@ -858,12 +860,10 @@ def helper_test_transfer(tag, ts, ts_new):
...
-    def helper_test_transfer_xa(tag, ts, xa_new):
-        """static cov or metadata must be identical between xarray and TimeSeries"""
-        if tag == STATIC_COV_TAG:
-            assert xa_new.attrs[STATIC_COV_TAG].equals(ts.static_covariates)
-        else:  # metadata
-            assert xa_new.attrs[METADATA_TAG] == ts.metadata
+    def helper_test_transfer_np(tag, ts, values):
+        """static cov and metadata are not transferred when performing value comparisons. Output is a `np.ndarray`"""
+        assert isinstance(values, np.ndarray)
+        assert values.shape == ts.shape
```
Comment (madtoinou) [inline darts/tests/test_timeseries_static_covariates.py:866]: This should probably be highlighted in the changelog as well

File: darts/timeseries.py [hunk 14]
```diff
@@ -2315,31 +2263,35 @@ def intvl(start, end):
...
     ) -> int:
-        """
-        Converts a point along the time axis index into an integer index ranging in (0, len(series)-1).
+        """Convert a point along the time index into an integer index ranging from (0, len(series)-1) inclusive.
 
         Parameters
         ----------
         point
-            This parameter supports 3 different data types: ``pd.Timestamp``, ``float`` and ``int``.
+            This parameter supports 3 different data types: ``pandas.Timestamp``, ``float`` and ``int``.
```
Comment (madtoinou) [inline darts/timeseries.py:2308]: Should we change all the `pd.` with `pandas.` in all the docstrings?

File: darts/timeseries.py [hunk 15]
```diff
@@ -2855,23 +2813,22 @@ def rescale_with_value(self, value_at_first_step: float) -> Self:
...
-            A new TimeSeries, where the first value is `value_at_first_step` and other values
-            have been scaled accordingly.
+            A new series, where the first value is `value_at_first_step` and other values have been scaled accordingly.
         """
-
-        raise_if_not(
-            (self._xa[0, :, :] != 0).all(), "Cannot rescale with first value 0.", logger
+        if (self._values[0, :, :] == 0).any():
+            raise_log(ValueError("Cannot rescale with first value `0`."), logger)
+        coef = value_at_first_step / self._values[:1]
```
Comment (madtoinou) [inline darts/timeseries.py:2837]: Great!

File: darts/timeseries.py [hunk 16]
```diff
@@ -5375,118 +5373,109 @@ def __truediv__(self, other):
...
+        return str(self.data_array(copy=False)).replace(
+            "xarray.DataArray", "TimeSeries (DataArray)"
+        )
 
     def __repr__(self):
-        return self._xa.__repr__().replace("xarray.DataArray", "TimeSeries (DataArray)")
+        return (
+            self.data_array(copy=False)
+            .__repr__()
+            .replace("xarray.DataArray", "TimeSeries (DataArray)")
```
Comment (madtoinou) [inline darts/timeseries.py]: Do we really want to keep the "(DataArray)" in the representation since it not used under the hood anymore?

**Review by madtoinou — COMMENTED** [review]
Incredible work @dennisbader!

A couple of remarks but the refactoring looks very good!

Can't wait to see it release.

File: CHANGELOG.md [hunk 17]
```diff
@@ -18,8 +18,8 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
   - 🟠 Renamed `RandomForest` to `RandomForestModel`. Using `RandomForest` will raise a depraction warning.
   - 🔴 Renamed `RegressionModelWithCategoricalCovariates` to `SKLearnModelWithCategoricalCovariates`. Removed `RegressionModelWithCategoricalCovariates`
 - Improvements to `TorchForecastingModel`: [#2802](https://github.com/unit8co/darts/pull/2802) by [Dennis Bader](https://github.com/dennisbader).
-  - Drastically improved prediction speed which is now up to 5.4 times as fast before. Using `n_jobs>1` now significantly boosts efficiency. This affects `predict()`, `historcial_forecasts()`, `backtest()`, `gridsearch()` and `residuals()`. The highest boost can be observed when calling historical forecasts, backtest, or residuals with `last_points_only=False`.
-  - Added parameter `values_only` to method `predict_from_dataset()` which will return a tuple of (prediction `np.ndarray`, target series schema, prediction start time) instead of `TimeSeries` objects.
+  - Drastically improved prediction speed which is now up to 5.4 times as fast before. This affects `predict()`, `historcial_forecasts()`, `backtest()`, `gridsearch()` and `residuals()`.
```
Comment (madtoinou) [inline CHANGELOG.md]:
```suggestion
  - Drastically improved prediction speed which is now up to 5.4 times as fast before. This affects `predict()`, `historical_forecasts()`, `backtest()`, `gridsearch()` and `residuals()`.
```

File: darts/timeseries.py (hunk 6 above)
Comment (dennisbader) [reply darts/timeseries.py]: In a separate issue. We need to thoroughly check that everything works as expected with tz info. I'm especially thinking about `n_steps_between` :D

File: darts/tests/test_timeseries.py (hunk 10 above)
Comment (dennisbader) [reply darts/tests/test_timeseries.py:60]: This is how it has always been for `from_times_and_values` and so on

File: darts/timeseries.py (hunk 16 above)
Comment (dennisbader) [reply darts/timeseries.py]: Removing it sounds good 👍
